tsl/src/chunk.c

```diff
@@ -421,3 +422,134 @@ chunk_drop_replica(PG_FUNCTION_ARGS)
 
 	PG_RETURN_VOID();
 }
+
+typedef struct
+{
+	Chunk *chunk;
+	NameData operation_id;
+	List *src_node;
+	List *dst_node;
+} ChunkCopyData;
+
+static void
+chunk_copy_data_prepare(ChunkCopyData *ccd)
+{
+	const char *cmd;
+	const char *connection_string;
+
+	/* create publication on the source data node */
+	cmd = psprintf("CREATE PUBLICATION %s FOR TABLE %s",
+				   NameStr(ccd->operation_id),
+				   quote_qualified_identifier(NameStr(ccd->chunk->fd.schema_name),
+											  NameStr(ccd->chunk->fd.table_name)));
+	ts_dist_cmd_run_on_data_nodes(cmd, ccd->src_node, false);
+
+	/* create subscription on the destination data node */
+
+	/*
+	 * CREATE SUBSCRIPTION from a database within the same database cluster will hang,
+	 * create the replication slot separately before creating the subscription
+	 */
+	cmd = psprintf("SELECT pg_create_logical_replication_slot('%s', 'pgoutput')",
+				   NameStr(ccd->operation_id));
+	ts_dist_cmd_run_on_data_nodes(cmd, ccd->src_node, false);
+
+	/* prepare connection string to the source node */
+	connection_string = remote_connection_get_connstr(lfirst(list_head(ccd->src_node)));
+
+	cmd = psprintf("CREATE SUBSCRIPTION %s CONNECTION '%s' PUBLICATION %s"
+				   " WITH (create_slot = false, enabled = false)",
+				   NameStr(ccd->operation_id),
+				   connection_string,
+				   NameStr(ccd->operation_id));
+	ts_dist_cmd_run_on_data_nodes(cmd, ccd->dst_node, false);
+}
+
+static void
+chunk_copy_data_perform(ChunkCopyData *ccd)
+{
+	const char *cmd;
+
+	/* start data transfer on the destination node */
+	cmd = psprintf("ALTER SUBSCRIPTION %s ENABLE", NameStr(ccd->operation_id));
+	ts_dist_cmd_run_on_data_nodes(cmd, ccd->dst_node, false);
+
+	/* wait until data transfer finishes */
+	cmd = psprintf("CALL _timescaledb_internal.wait_subscription_sync(%s, %s)",
```

Comment:
This will create an infinite loop wait because you are creating SUBSCRIPTION in the same transaction, so there won't be an entry that is visible in the `pg_subscription_rel` catalog yet. The wait has to happen after the CREATE SUBS/ALTER ENABLE has committed..

Reply:
I guess we figured it out during discussions :)